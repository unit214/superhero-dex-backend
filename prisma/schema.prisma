// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Token {
  id         Int     @id @default(autoincrement())
  address    String  @unique
  symbol     String  @default("n/a")
  name       String  @default("n/a")
  decimals   Int     @default(-1)
  malformed  Boolean @default(false)
  noContract Boolean @default(false)
  pairs0     Pair[]  @relation("Pair_Token0")
  pairs1     Pair[]  @relation("Pair_Token1")
  listed     Boolean @default(false)
}

model Pair {
  id                          Int                               @id @default(autoincrement())
  address                     String                            @unique
  token0                      Token                             @relation("Pair_Token0", fields: [t0], references: [id])
  t0                          Int
  token1                      Token                             @relation("Pair_Token1", fields: [t1], references: [id])
  t1                          Int
  liquidityInfo               PairLiquidityInfo?                @relation("Pair_Info")
  synchronized                Boolean
  liquidityInfoHistory        PairLiquidityInfoHistory[]
  liquidityInfoHistoryV2      PairLiquidityInfoHistoryV2[]
  liquidityInfoHistoryError   PairLiquidityInfoHistoryError[]
  liquidityInfoHistoryV2Error PairLiquidityInfoHistoryV2Error[]
}

model PairLiquidityInfo {
  id          Int    @id
  pair        Pair   @relation("Pair_Info", fields: [id], references: [id])
  totalSupply String
  reserve0    String
  reserve1    String
  height      Int
}

model PairLiquidityInfoHistory {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  totalSupply    String
  reserve0       String
  reserve1       String
  height         Int
  microBlockHash String
  microBlockTime BigInt
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashUniqueIndex", [pairId, microBlockHash])
}

model PairLiquidityInfoHistoryError {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashErrorUniqueIndex", [pairId, microBlockHash, error])
}

model PairLiquidityInfoHistoryV2 {
  id              Int      @id @default(autoincrement())
  pair            Pair     @relation(fields: [pairId], references: [id])
  pairId          Int
  eventType       String
  height          Int
  microBlockTime  BigInt
  logIndex        Int
  microBlockHash  String
  transactionHash String
  totalSupply     String?
  reserve0        String?
  reserve1        String?
  deltaReserve0   String?
  deltaReserve1   String?
  fiatPrice       String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex])
}

model PairLiquidityInfoHistoryV2Error {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  logIndex       Int
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex, error])
}
