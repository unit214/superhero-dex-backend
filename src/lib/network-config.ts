type NetworkConfig = {
  nodeUrl: string;
  compilerUrl: string;
  middlewareWebsocketUrl: string;
  middlewareHttpUrl: string;
};

export default {
  testnet: {
    nodeUrl: 'https://testnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareWebsocketUrl: 'wss://testnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://testnet.aeternity.io/mdw',
  },
  mainnet: {
    nodeUrl: 'https://mainnet.aeternity.io',
    compilerUrl: 'https://latest.compiler.aepps.com',
    middlewareWebsocketUrl: 'wss://mainnet.aeternity.io/mdw/v2/websocket',
    middlewareHttpUrl: 'https://mainnet.aeternity.io/mdw',
  },
} as { [key: string]: NetworkConfig };
