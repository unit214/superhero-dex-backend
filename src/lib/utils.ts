import { Decimal } from '@prisma/client/runtime/library';
import BigNumber from 'bignumber.js';

export const nonNullable = <T>(t: T | null | undefined, label?: string): T => {
  if (t == null) {
    throw new Error(
      `${label ? label + ': ' : null} non nullable value expected`,
    );
  }
  return t;
};

export const removeId = <ID, T extends { id: ID }>(t: T) => {
  const {
    id, // eslint-disable-line @typescript-eslint/no-unused-vars
    ...tail
  } = t;
  return tail;
};

export const pluralize = (count: number, noun: string, suffix = 's') =>
  `${count} ${noun}${count !== 1 ? suffix : ''}`;

const parseEnv = (x) => x && JSON.parse(x);
export const presentInvalidTokens = parseEnv(process.env.SHOW_INVALID_TOKENS);

export const numberToDecimal = (number: number): Decimal =>
  new Decimal(number.toString());

export const bigIntToDecimal = (bigInt: bigint): Decimal =>
  new Decimal(bigInt.toString());

export const decimalToBigInt = (decimal: Decimal): bigint =>
  BigInt(decimal.toFixed().toString());

export const calculateUsdValue = ({
  reserve,
  tokenAePrice,
  decimals,
  aeUsdPrice,
}: {
  reserve: string;
  tokenAePrice: string;
  decimals: number;
  aeUsdPrice: string;
}) => {
  if (!reserve || !tokenAePrice || !aeUsdPrice) {
    return '0';
  }

  return new BigNumber(reserve)
    .abs()
    .div(new BigNumber(10).pow(decimals))
    .multipliedBy(tokenAePrice)
    .multipliedBy(aeUsdPrice)
    .toString();
};
