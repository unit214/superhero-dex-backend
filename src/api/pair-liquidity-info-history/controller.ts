import {
  Controller,
  Get,
  ParseEnumPipe,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { PairLiquidityInfoHistoryService } from './service';
import { ApiOperation, ApiQuery, ApiResponse } from '@nestjs/swagger';
import * as dto from '../../dto';
import { OrderQueryEnum } from '../../dto';
import { ContractAddress } from '../../lib/utils';

@Controller('history/liquidity')
export class PairLiquidityInfoHistoryController {
  constructor(
    private readonly pairLiquidityInfoHistoryService: PairLiquidityInfoHistoryService,
  ) {}

  @Get()
  @ApiOperation({
    summary: 'Retrieve all entries of the pair liquidity info history',
    deprecated: true,
  })
  @ApiQuery({
    name: 'limit',
    type: Number,
    description: 'Limit of history entries per page (default: 100, max: 100)',
    required: false,
  })
  @ApiQuery({
    name: 'offset',
    type: Number,
    description: 'Offset of page (default: 0)',
    required: false,
  })
  @ApiQuery({
    name: 'order',
    enum: OrderQueryEnum,
    description:
      'Sorts history entries in ascending or descending order (default: asc)',
    required: false,
  })
  @ApiQuery({
    name: 'pair-address',
    type: String,
    description: 'Retrieve only history entries for the given pair address',
    required: false,
  })
  @ApiQuery({
    name: 'height',
    type: Number,
    description: 'Retrieve only history entries for the given height',
    required: false,
  })
  @ApiQuery({
    name: 'from-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or newer than the given micro block time',
    required: false,
  })
  @ApiQuery({
    name: 'to-block-time',
    type: Number,
    description:
      'Retrieve only history entries that are equal or older than the given micro block time',
    required: false,
  })
  @ApiResponse({ status: 200, type: [dto.PairLiquidityInfoHistoryEntry] })
  async findAll(
    @Query('limit', new ParseIntPipe({ optional: true })) limit: number = 100,
    @Query('offset', new ParseIntPipe({ optional: true })) offset: number = 0,
    @Query('order', new ParseEnumPipe(OrderQueryEnum, { optional: true }))
    order: OrderQueryEnum = OrderQueryEnum.asc,
    @Query('pair-address') pairAddress?: ContractAddress,
    @Query('height', new ParseIntPipe({ optional: true })) height?: number,
    @Query('from-block-time', new ParseIntPipe({ optional: true }))
    fromBlockTime?: number,
    @Query('to-block-time', new ParseIntPipe({ optional: true }))
    toBlockTime?: number,
  ): Promise<dto.PairLiquidityInfoHistoryEntry[]> {
    return this.pairLiquidityInfoHistoryService
      .getAllHistoryEntries(
        Number(limit),
        Number(offset),
        order,
        pairAddress,
        height ? Number(height) : undefined,
        fromBlockTime ? BigInt(fromBlockTime) : undefined,
        toBlockTime ? BigInt(toBlockTime) : undefined,
      )
      .then((entries) =>
        entries.map((entry) => ({
          pairAddress: entry.pair.address,
          liquidityInfo: {
            totalSupply: entry.totalSupply,
            reserve0: entry.reserve0,
            reserve1: entry.reserve1,
          },
          height: entry.height,
          microBlockHash: entry.microBlockHash,
          microBlockTime: entry.microBlockTime.toString(),
        })),
      );
  }
}
