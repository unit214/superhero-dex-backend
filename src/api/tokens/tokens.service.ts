import { Cache } from '@nestjs/cache-manager';
import { Injectable } from '@nestjs/common';
import { Pair, PairLiquidityInfo, Token } from '@prisma/client';

import { TokenWithUsd } from '@/api/api.model';
import { ContractAddress } from '@/clients/sdk-client.model';
import { TokenDbService } from '@/database/token/token-db.service';
import { presentInvalidTokens } from '@/lib/utils';

@Injectable()
export class TokensService {
  constructor(
    private readonly tokenDbService: TokenDbService,
    private cacheManager: Cache,
  ) {}
  async getCount(onlyListed?: boolean) {
    return this.tokenDbService.count(presentInvalidTokens, onlyListed);
  }
  async getAllTokens(): Promise<Token[]> {
    return this.tokenDbService.getAll(presentInvalidTokens);
  }
  async getAllTokensWithAggregation(): Promise<TokenWithUsd[]> {
    return this.tokenDbService.getAllWithAggregation(presentInvalidTokens);
  }
  async getListedTokens(): Promise<Token[]> {
    return this.tokenDbService.getListed();
  }

  async listToken(address: ContractAddress) {
    const res = this.tokenDbService.updateListedValue(address, true);
    await this.cacheManager.reset();
    return res;
  }

  async unlistToken(address: ContractAddress) {
    const res = this.tokenDbService.updateListedValue(address, false);
    await this.cacheManager.reset();
    return res;
  }

  async getToken(address: ContractAddress): Promise<TokenWithUsd | null> {
    return this.tokenDbService.getWithAggregation(address);
  }

  async getTokenWithPairsInfo(address: ContractAddress): Promise<
    | (Token & {
        pairs0: (Pair & {
          token1: Token;
          liquidityInfo: PairLiquidityInfo | null;
        })[];
        pairs1: (Pair & {
          liquidityInfo: PairLiquidityInfo | null;
          token0: Token;
        })[];
      })
    | null
  > {
    return this.tokenDbService.getByAddressWithPairsAndLiquidity(address);
  }
}
