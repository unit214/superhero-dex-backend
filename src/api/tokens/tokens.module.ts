import { Module } from '@nestjs/common';
import { TokensController } from './tokens.controller';
import { TokensService } from './tokens.service';

@Module({
  imports: [],
  controllers: [TokensController],
  providers: [TokensService],
})
export class TokensModule {}
