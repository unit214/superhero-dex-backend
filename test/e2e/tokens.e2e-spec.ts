import * as process from 'node:process';

import { CacheModule } from '@nestjs/cache-manager';
import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';

import { TokensController } from '@/api/tokens/tokens.controller';
import { TokensService } from '@/api/tokens/tokens.service';
import { PrismaService } from '@/database/prisma.service';
import { TokenDbService } from '@/database/token/token-db.service';
import { nonNullable } from '@/lib/utils';
import {
  historyEntry1,
  historyEntry2,
  historyEntry3,
  historyEntry4,
  liquidityInfo1,
  liquidityInfo2,
  pair1,
  pair2,
  pair3,
  token1,
  token2,
  token3,
  token5,
} from '@/test/mock-data/pair-liquidity-info-history-mock-data';

// Testing method
// before all
//   - initiate nest app
// before each
//   - clean db
//   - insert test data
// after all
//   - close nest app
//   - disconnect from prisma
describe('TokenController', () => {
  let app: INestApplication;
  let prismaService: PrismaService;

  const authToken = nonNullable(process.env.AUTH_TOKEN);

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [TokensController],
      imports: [CacheModule.register({})],
      providers: [TokenDbService, TokensService, PrismaService],
    }).compile();

    app = module.createNestApplication();
    await app.init();

    prismaService = module.get(PrismaService);
  });

  beforeAll(async () => {
    await prismaService.pairLiquidityInfoHistory.deleteMany();
    await prismaService.pairLiquidityInfo.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  beforeEach(async () => {
    await prismaService.token.createMany({
      data: [token1, token2, token3, token5],
    });
    await prismaService.pair.createMany({ data: [pair1, pair2, pair3] });
    await prismaService.pairLiquidityInfo.createMany({
      data: [liquidityInfo1, liquidityInfo2],
    });
    await prismaService.pairLiquidityInfoHistory.createMany({
      data: [historyEntry1, historyEntry2, historyEntry3, historyEntry4],
    });
  });

  afterEach(async () => {
    await prismaService.pairLiquidityInfoHistory.deleteMany();
    await prismaService.pairLiquidityInfo.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  afterAll(async () => {
    await prismaService.$disconnect();
    await app.close();
  });

  describe('GET /tokens', () => {
    it('should return all tokens when none are listed', async () => {
      await request(app.getHttpServer())
        .get('/tokens')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return all tokens even if some are listed', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .set({ Authorization: authToken });
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token3')
        .set({ Authorization: authToken });
      await request(app.getHttpServer())
        .get('/tokens')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });
  });

  describe('GET /tokens/{token_address}', () => {
    it('should return a single token if it exists', async () => {
      await request(app.getHttpServer())
        .get('/tokens/ct_token1')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return 404 if the token does not exist', () => {
      return request(app.getHttpServer())
        .get('/tokens/ct_xxxx')
        .expect(404)
        .expect({
          statusCode: 404,
          message: 'token not found',
          error: 'Not Found',
        });
    });
  });

  describe('GET /tokens/{token_address}/pairs', () => {
    it('should return pairs with no liquidity info', async () => {
      await request(app.getHttpServer())
        .get('/tokens/ct_token1/pairs')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return pairs with liquidity info', async () => {
      await request(app.getHttpServer())
        .get('/tokens/ct_token2/pairs')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });

    it('should return 404 if the token does not exist', () => {
      return request(app.getHttpServer())
        .get('/tokens/ct_xxxx')
        .expect(404)
        .expect({
          statusCode: 404,
          message: 'token not found',
          error: 'Not Found',
        });
    });
  });

  describe('GET /tokens/listed', () => {
    it('should return no tokens if none are listed', () => {
      return request(app.getHttpServer())
        .get('/tokens/listed')
        .expect(200)
        .expect([]);
    });

    it('should return listed tokens if some are listed', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .set('Authorization', authToken);
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token3')
        .set('Authorization', authToken);
      await request(app.getHttpServer())
        .get('/tokens/listed')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });
  });

  describe('POST /tokens/listed/{token_address}', () => {
    it('should return 401 with no auth key and invalid token address', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_xxxx')
        .expect(401);
    });

    it('should return 401 with no auth key and valid token address', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .expect(401);
    });

    it('should return 401 with invalid auth key and invalid token', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_xxxx')
        .set('Authorization', 'wrong-key')
        .expect(401);
    });

    it('should return 401 with invalid auth key and valid token', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .set('Authorization', 'wrong-key')
        .expect(401);
    });

    it('should return 404 with valid auth key but with invalid token', async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_xxxx')
        .set('Authorization', authToken)
        .expect(404);
    });

    it('should return 201 with valid auth key and valid token and mark the token as listed', async () => {
      //verify before listing ct_t1
      await request(app.getHttpServer())
        .get('/tokens/ct_token1')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());

      //listing it
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token1')
        .set('Authorization', authToken)
        .expect(201)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());

      //re-verify ct_t1 to be sure it was persisted also
      await request(app.getHttpServer())
        .get('/tokens/ct_token1')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });
  });

  describe('DELETE /tokens/listed/{token_address}', () => {
    beforeEach(async () => {
      await request(app.getHttpServer())
        .post('/tokens/listed/ct_token3')
        .set('Authorization', authToken);
    });

    it('should return 401 with no auth key and invalid token', async () => {
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_xxxx')
        .expect(401);
    });

    it('should return 401 with no auth key and valid token address', async () => {
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_token3')
        .expect(401);
    });

    it('should return 401 with invalid auth key and invalid token', async () => {
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_xxxx')
        .set('Authorization', 'wrong-key')
        .expect(401);
    });

    it('should return 401 with invalid auth key and valid token', async () => {
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_token3')
        .set('Authorization', 'wrong-key')
        .expect(401);
    });

    it('should return 404 with valid auth key but with invalid token', async () => {
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_xxxx')
        .set('Authorization', authToken)
        .expect(404);
    });

    it('should return 200 with valid auth key and valid token', async () => {
      //verify before unlisting ct_t3
      await request(app.getHttpServer())
        .get('/tokens/ct_token3')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());

      //unlisting it
      await request(app.getHttpServer())
        .delete('/tokens/listed/ct_token3')
        .set('Authorization', nonNullable(process.env.AUTH_TOKEN))
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());

      //re-verify ct_t3 to be sure the unlisting was persisted too
      await request(app.getHttpServer())
        .get('/tokens/ct_token3')
        .expect(200)
        .then((res) => expect(JSON.parse(res.text)).toMatchSnapshot());
    });
  });
});
