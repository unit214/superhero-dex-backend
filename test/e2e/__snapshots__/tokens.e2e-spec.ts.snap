// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenController DELETE /tokens/listed/{token_address} should return 200 with valid auth key and valid token 1`] = `
{
  "address": "ct_token3",
  "decimals": 18,
  "listed": true,
  "malformed": false,
  "name": "3",
  "noContract": false,
  "pairs": 1,
  "priceAe": null,
  "priceChangeDay": null,
  "priceChangeMonth": null,
  "priceChangeWeek": null,
  "priceChangeYear": null,
  "priceUsd": null,
  "symbol": "3",
  "totalReserve": null,
  "tvlAe": "4.8e-17",
  "tvlUsd": "0",
  "volumeUsdAll": "0",
  "volumeUsdDay": "0",
  "volumeUsdMonth": "0",
  "volumeUsdWeek": "0",
  "volumeUsdYear": "0",
}
`;

exports[`TokenController DELETE /tokens/listed/{token_address} should return 200 with valid auth key and valid token 2`] = `
{
  "address": "ct_token3",
  "decimals": 18,
  "listed": false,
  "malformed": false,
  "name": "3",
  "noContract": false,
  "symbol": "3",
}
`;

exports[`TokenController DELETE /tokens/listed/{token_address} should return 200 with valid auth key and valid token 3`] = `
{
  "address": "ct_token3",
  "decimals": 18,
  "listed": false,
  "malformed": false,
  "name": "3",
  "noContract": false,
  "pairs": 1,
  "priceAe": null,
  "priceChangeDay": null,
  "priceChangeMonth": null,
  "priceChangeWeek": null,
  "priceChangeYear": null,
  "priceUsd": null,
  "symbol": "3",
  "totalReserve": null,
  "tvlAe": "4.8e-17",
  "tvlUsd": "0",
  "volumeUsdAll": "0",
  "volumeUsdDay": "0",
  "volumeUsdMonth": "0",
  "volumeUsdWeek": "0",
  "volumeUsdYear": "0",
}
`;

exports[`TokenController GET /tokens should return all tokens even if some are listed 1`] = `
[
  {
    "address": "ct_token1",
    "decimals": 18,
    "listed": true,
    "malformed": false,
    "name": "1",
    "noContract": false,
    "pairs": 1,
    "priceAe": "0.05308695",
    "priceChangeDay": "0",
    "priceChangeMonth": "0",
    "priceChangeWeek": "0",
    "priceChangeYear": "0",
    "priceUsd": "0.0027",
    "symbol": "1",
    "totalReserve": "1e-15",
    "tvlAe": "5.3e-17",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token2",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "2",
    "noContract": false,
    "pairs": 3,
    "priceAe": "0.101118",
    "priceChangeDay": "0",
    "priceChangeMonth": "0",
    "priceChangeWeek": "0",
    "priceChangeYear": "0",
    "priceUsd": "0.0051",
    "symbol": "2",
    "totalReserve": "1e-15",
    "tvlAe": "1.01e-16",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token3",
    "decimals": 18,
    "listed": true,
    "malformed": false,
    "name": "3",
    "noContract": false,
    "pairs": 1,
    "priceAe": null,
    "priceChangeDay": null,
    "priceChangeMonth": null,
    "priceChangeWeek": null,
    "priceChangeYear": null,
    "priceUsd": null,
    "symbol": "3",
    "totalReserve": null,
    "tvlAe": "4.8e-17",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token5",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "5",
    "noContract": false,
    "pairs": 1,
    "priceAe": null,
    "priceChangeDay": null,
    "priceChangeMonth": null,
    "priceChangeWeek": null,
    "priceChangeYear": null,
    "priceUsd": null,
    "symbol": "5",
    "totalReserve": null,
    "tvlAe": null,
    "tvlUsd": null,
    "volumeUsdAll": null,
    "volumeUsdDay": null,
    "volumeUsdMonth": null,
    "volumeUsdWeek": null,
    "volumeUsdYear": null,
  },
]
`;

exports[`TokenController GET /tokens should return all tokens when none are listed 1`] = `
[
  {
    "address": "ct_token1",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "1",
    "noContract": false,
    "pairs": 1,
    "priceAe": "0.05308695",
    "priceChangeDay": "0",
    "priceChangeMonth": "0",
    "priceChangeWeek": "0",
    "priceChangeYear": "0",
    "priceUsd": "0.0027",
    "symbol": "1",
    "totalReserve": "1e-15",
    "tvlAe": "5.3e-17",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token2",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "2",
    "noContract": false,
    "pairs": 3,
    "priceAe": "0.101118",
    "priceChangeDay": "0",
    "priceChangeMonth": "0",
    "priceChangeWeek": "0",
    "priceChangeYear": "0",
    "priceUsd": "0.0051",
    "symbol": "2",
    "totalReserve": "1e-15",
    "tvlAe": "1.01e-16",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token3",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "3",
    "noContract": false,
    "pairs": 1,
    "priceAe": null,
    "priceChangeDay": null,
    "priceChangeMonth": null,
    "priceChangeWeek": null,
    "priceChangeYear": null,
    "priceUsd": null,
    "symbol": "3",
    "totalReserve": null,
    "tvlAe": "4.8e-17",
    "tvlUsd": "0",
    "volumeUsdAll": "0",
    "volumeUsdDay": "0",
    "volumeUsdMonth": "0",
    "volumeUsdWeek": "0",
    "volumeUsdYear": "0",
  },
  {
    "address": "ct_token5",
    "decimals": 18,
    "listed": false,
    "malformed": false,
    "name": "5",
    "noContract": false,
    "pairs": 1,
    "priceAe": null,
    "priceChangeDay": null,
    "priceChangeMonth": null,
    "priceChangeWeek": null,
    "priceChangeYear": null,
    "priceUsd": null,
    "symbol": "5",
    "totalReserve": null,
    "tvlAe": null,
    "tvlUsd": null,
    "volumeUsdAll": null,
    "volumeUsdDay": null,
    "volumeUsdMonth": null,
    "volumeUsdWeek": null,
    "volumeUsdYear": null,
  },
]
`;

exports[`TokenController GET /tokens/{token_address} should return a single token if it exists 1`] = `
{
  "address": "ct_token1",
  "decimals": 18,
  "listed": false,
  "malformed": false,
  "name": "1",
  "noContract": false,
  "pairs": 1,
  "priceAe": "0.05308695",
  "priceChangeDay": "0",
  "priceChangeMonth": "0",
  "priceChangeWeek": "0",
  "priceChangeYear": "0",
  "priceUsd": "0.0027",
  "symbol": "1",
  "totalReserve": "1e-15",
  "tvlAe": "5.3e-17",
  "tvlUsd": "0",
  "volumeUsdAll": "0",
  "volumeUsdDay": "0",
  "volumeUsdMonth": "0",
  "volumeUsdWeek": "0",
  "volumeUsdYear": "0",
}
`;

exports[`TokenController GET /tokens/{token_address}/pairs should return pairs with liquidity info 1`] = `
{
  "pairs0": [
    {
      "address": "ct_pair2",
      "liquidityInfo": {
        "height": 1,
        "reserve0": "1",
        "reserve1": "1",
        "totalSupply": "1",
      },
      "oppositeToken": {
        "address": "ct_token3",
        "decimals": 18,
        "listed": false,
        "malformed": false,
        "name": "3",
        "noContract": false,
        "symbol": "3",
      },
      "synchronized": true,
    },
    {
      "address": "ct_pair3",
      "liquidityInfo": {
        "height": 2,
        "reserve0": "2",
        "reserve1": "2",
        "totalSupply": "2",
      },
      "oppositeToken": {
        "address": "ct_token5",
        "decimals": 18,
        "listed": false,
        "malformed": false,
        "name": "5",
        "noContract": false,
        "symbol": "5",
      },
      "synchronized": true,
    },
  ],
  "pairs1": [
    {
      "address": "ct_pair1",
      "oppositeToken": {
        "address": "ct_token1",
        "decimals": 18,
        "listed": false,
        "malformed": false,
        "name": "1",
        "noContract": false,
        "symbol": "1",
      },
      "synchronized": true,
    },
  ],
}
`;

exports[`TokenController GET /tokens/{token_address}/pairs should return pairs with no liquidity info 1`] = `
{
  "pairs0": [
    {
      "address": "ct_pair1",
      "oppositeToken": {
        "address": "ct_token2",
        "decimals": 18,
        "listed": false,
        "malformed": false,
        "name": "2",
        "noContract": false,
        "symbol": "2",
      },
      "synchronized": true,
    },
  ],
  "pairs1": [],
}
`;

exports[`TokenController GET /tokens/listed should return listed tokens if some are listed 1`] = `
[
  {
    "address": "ct_token1",
    "decimals": 18,
    "malformed": false,
    "name": "1",
    "noContract": false,
    "symbol": "1",
  },
  {
    "address": "ct_token3",
    "decimals": 18,
    "malformed": false,
    "name": "3",
    "noContract": false,
    "symbol": "3",
  },
]
`;

exports[`TokenController POST /tokens/listed/{token_address} should return 201 with valid auth key and valid token and mark the token as listed 1`] = `
{
  "address": "ct_token1",
  "decimals": 18,
  "listed": false,
  "malformed": false,
  "name": "1",
  "noContract": false,
  "pairs": 1,
  "priceAe": "0.05308695",
  "priceChangeDay": "0",
  "priceChangeMonth": "0",
  "priceChangeWeek": "0",
  "priceChangeYear": "0",
  "priceUsd": "0.0027",
  "symbol": "1",
  "totalReserve": "1e-15",
  "tvlAe": "5.3e-17",
  "tvlUsd": "0",
  "volumeUsdAll": "0",
  "volumeUsdDay": "0",
  "volumeUsdMonth": "0",
  "volumeUsdWeek": "0",
  "volumeUsdYear": "0",
}
`;

exports[`TokenController POST /tokens/listed/{token_address} should return 201 with valid auth key and valid token and mark the token as listed 2`] = `
{
  "address": "ct_token1",
  "decimals": 18,
  "listed": true,
  "malformed": false,
  "name": "1",
  "noContract": false,
  "symbol": "1",
}
`;

exports[`TokenController POST /tokens/listed/{token_address} should return 201 with valid auth key and valid token and mark the token as listed 3`] = `
{
  "address": "ct_token1",
  "decimals": 18,
  "listed": true,
  "malformed": false,
  "name": "1",
  "noContract": false,
  "pairs": 1,
  "priceAe": "0.05308695",
  "priceChangeDay": "0",
  "priceChangeMonth": "0",
  "priceChangeWeek": "0",
  "priceChangeYear": "0",
  "priceUsd": "0.0027",
  "symbol": "1",
  "totalReserve": "1e-15",
  "tvlAe": "5.3e-17",
  "tvlUsd": "0",
  "volumeUsdAll": "0",
  "volumeUsdDay": "0",
  "volumeUsdMonth": "0",
  "volumeUsdWeek": "0",
  "volumeUsdYear": "0",
}
`;
